File: app/frontend/src/api/fragment.ts
// 创作流碎片：作者自己记下的东西（灵感速记 / 事件 / 口述段落）。
//
// 这一层只做"取值 ↔ 参数"的转换；种类闭集、软删与捞回、留痕全在核心。
//
// 三个口径（与核心一致，界面上也照这么说）：
// - **与叩问同住一张表**：问题卡、答案、灵感、事件是一条存储上的不同种类——
//   所以它们共享同一套元数据（怎么记下的 / 多新鲜 / 用没用过 / 关联着哪一章）；
// - **记下来不打断**：写一条碎片没有别的副作用（不动问题状态、不动正文）；
// - **删只是打时间戳**：捞回就是把时间戳抹掉，所以"撤销"是能兑现的。

import { call, COMMANDS } from "./core";

/** 碎片的种类（与核心 `FragmentKind` 的稳定码一致）。 */
export type FragmentKind = "question" | "answer" | "idea" | "event" | "dictation";

/** 一条碎片（面板读到的统一形状）。 */
export interface Fragment {
  id: number;
  work_id: number;
  kind: FragmentKind;
  body: string;
  /** 怎么记下的：`typed` / `voice` / `mixed`（与答案同一列、同一闭集）。 */
  source: string;
  /** 关联锚点（`chapter:12` 这种）。 */
  anchors: string[];
  /** 从哪张问题卡勾出来的（自己随手记的没有）。 */
  derived_from: number | null;
  created_at: number;
}

/** 一种碎片有几条（面板上那些筛选项的数字）。 */
export interface FragmentCount {
  kind: FragmentKind;
  count: number;
}

/** 创作流面板一屏要的全部数据（列表 + 各筛选项的数字，一次给全免得前后对不上）。 */
export interface CreatorBoard {
  fragments: Fragment[];
  counts: FragmentCount[];
}

/**
 * 记一条碎片。
 *
 * `kind` 只认作者自己记的那几种（`idea` / `event` / `dictation`）——问题卡与答案归叩问，
 * 从这儿建会被核心当场拒（`fragment.kind_not_jotted`）。
 * `anchors` 是关联锚点：面板把「是在这一章写的」记成 `chapter:<章 id>`。
 */
export const fragmentAdd = (
  work_id: number,
  kind: FragmentKind,
  body: string,
  source: string,
  anchors: string[] = [],
) => call<Fragment>(COMMANDS.fragmentAdd, { work_id, kind, body, source, anchors });

/** 看一眼碎片池（不写库）。 */
export const fragmentBoard = (work_id: number) =>
  call<CreatorBoard>(COMMANDS.fragmentBoard, { work_id });

/** 删掉一条（**软删**：库里还留着，随时能捞回）。 */
export const fragmentDelete = (id: number) =>
  call<CreatorBoard>(COMMANDS.fragmentDelete, { id });

/** 捞回刚删掉的那一条（重复点也不算错）。 */
export const fragmentRestore = (id: number) =>
  call<CreatorBoard>(COMMANDS.fragmentRestore, { id });

File: app/src/commands/fragment.rs
//! 创作流命令域：**你自己记，研墨只替你归位**——碎片池的记 / 看 / 删 / 捞回。
//!
//! 与叩问那条线的分工：问题卡、答案、处置四件套归 [`crate::commands::question`]；
//! 这里只管作者随手记下的那几种（灵感速记 / 事件 / 口述段落），核心会挡着不让
//! 从这里建问题卡或答案（`fragment.kind_not_jotted`）。
//!
//! 面板的数据形状是**一次给全**（列表 + 各筛选项的数字）：它们是同一屏上的同一件事，
//! 分两次问就会让"全部 3 条"和列表里的两条对不上。

use serde::Serialize;
use tauri::State;

use crate::error::ApiError;
use crate::storage::AppData;
use yanmo_core::model::{Fragment, FragmentCount, FragmentKind};
use yanmo_core::store::{NewFragment, Store, FRAGMENTS_PER_BOARD};

/// 创作流面板一次要的全部数据。
#[derive(Debug, Serialize)]
pub struct CreatorBoardDto {
    /// 最近记下的那些（**新的在前**），作者记的那几种
    pub fragments: Vec<Fragment>,
    /// 每一种各有几条（**只数没删的**）——面板上那些筛选项的数字
    pub counts: Vec<FragmentCount>,
}

/// 面板读一次。种类固定是"作者能随手记"那一组：问题与答案有自己的处置面板，
/// 混进来只会让数字和列表各说各话。
fn board(store: &Store, work_id: i64) -> yanmo_core::Result<CreatorBoardDto> {
    let kinds = FragmentKind::JOTTED;
    Ok(CreatorBoardDto {
        fragments: store.fragments(work_id, &kinds, FRAGMENTS_PER_BOARD)?,
        counts: store.fragment_counts(work_id, &kinds)?,
    })
}

/// 记一条碎片。
///
/// `kind` 只认作者自己记的那几种（`idea` / `event` / `dictation`）；`source` 是
/// **怎么记下的**（`typed` / `voice` / `mixed`，闭集，认不出就拒）。`anchors` 是关联锚点
/// （界面把"是在这一章写的"记成 `chapter:<章 id>`）——核心只存不懂它，怎么用它归后面那条线。
#[tauri::command(rename_all = "snake_case")]
pub fn fragment_add(
    data: State<'_, AppData>,
    work_id: i64,
    kind: String,
    body: String,
    source: String,
    anchors: Vec<String>,
) -> Result<Fragment, ApiError> {
    crate::acceptance::note_command("fragment_add");
    data.with_store(|store| {
        let kind = FragmentKind::parse(&kind)?;
        let id = store.create_fragment(
            &NewFragment {
                work_id,
                kind,
                body,
                source,
                anchors,
            },
            "author",
        )?;
        // 回执给**库里真有的那一条**（修剪过的原文 + 核心认下的输入方式），不是界面自己回显
        store.fragment(id)
    })
}

/// 看一眼碎片池（不写库）。
#[tauri::command(rename_all = "snake_case")]
pub fn fragment_board(data: State<'_, AppData>, work_id: i64) -> Result<CreatorBoardDto, ApiError> {
    crate::acceptance::note_command("fragment_board");
    data.with_store(|store| board(store, work_id))
}

/// 删掉一条（**软删**：库里还留着，`fragment_restore` 能捞回来）。
///
/// 回执是删完之后的最新一屏——界面拿它直接换掉手上那份，不必自己猜少了哪一条。
#[tauri::command(rename_all = "snake_case")]
pub fn fragment_delete(data: State<'_, AppData>, id: i64) -> Result<CreatorBoardDto, ApiError> {
    crate::acceptance::note_command("fragment_delete");
    data.with_store(|store| {
        // 回执是它删掉的那一条（软删之后按"取一条"就查不到了），拿它的 work_id 回一屏最新的
        let gone = store.delete_fragment(id, "author")?;
        board(store, gone.work_id)
    })
}

/// 捞回一条删掉的（重复点不算错：已经在的原样返回）。
#[tauri::command(rename_all = "snake_case")]
pub fn fragment_restore(data: State<'_, AppData>, id: i64) -> Result<CreatorBoardDto, ApiError> {
    crate::acceptance::note_command("fragment_restore");
    data.with_store(|store| {
        let back = store.restore_fragment(id, "author")?;
        board(store, back.work_id)
    })
}
